core/services/workflows/monitoring.go

```diff
@@ -0,0 +1,227 @@
+package workflows
+
+import (
+	"context"
+	"fmt"
+	"github.com/smartcontractkit/chainlink-common/pkg/beholder"
+	"github.com/smartcontractkit/chainlink-common/pkg/beholder/pb/keystone"
+	"go.opentelemetry.io/otel/metric"
+	"google.golang.org/protobuf/proto"
+)
+
+type customMessageLabeler struct {
+	labels map[string]string
+}
+
+func NewCustomMessageLabeler() customMessageLabeler {
+	return customMessageLabeler{labels: make(map[string]string)}
+}
+
+// with adds multiple key-value pairs to the customMessageLabeler for transmission with sendLogAsCustomMessage
+func (c customMessageLabeler) with(keyValues ...string) customMessageLabeler {
+	newCustomMessageLabeler := NewCustomMessageLabeler()
+
+	if len(keyValues)%2 != 0 {
+		// If an odd number of key-value arguments is passed, return the original customMessageLabeler unchanged
+		return c
+	}
+
+	// Copy existing labels from the current agent
+	for k, v := range c.labels {
+		newCustomMessageLabeler.labels[k] = v
+	}
+
+	// Add new key-value pairs
+	for i := 0; i < len(keyValues); i += 2 {
+		key := keyValues[i]
+		value := keyValues[i+1]
+		newCustomMessageLabeler.labels[key] = value
+	}
+
+	return newCustomMessageLabeler
+}
+
+// sendLogAsCustomMessage emits a KeystoneCustomMessage with msg and labels as data.
+// any key in labels that is not part of orderedLabelKeys will not be transmitted
+func (c customMessageLabeler) sendLogAsCustomMessage(msg string) error {
```

Comment:
/nit could consider taking args ~ fmt string, args ...any 

and doing the fmt.Sprintf(fmt, args) on behalf of the caller.

not worth refactoring atm.